### PR TITLE
Add ability to skip directories

### PR DESCRIPTION
**Pull request by makubacki**
Adds a GetSkippedDirectories() method that allows a SettingsManager
to specify a tuple of workspace-relative directories that should be
skipped.

This allows subdirectories (such as submodules) that might contain duplicate
resources that would normally cause build errors (such as external dependency
descriptors) to be ignored if only one directory is actually being used for
the current build.

### CODE REVIEW

File: edk2toolext/environment/self_describing_environment.py [hunk 1]
```diff
@@ -46,8 +52,8 @@ def _gather_env_files(self, ext_strings, base_path):
         # matching the extension.
         matches = {}
         for root, dirs, files in os.walk(base_path, topdown=True):
-            # Check to see whether any of these directories should be skipped.
-            # TODO: Allow these to be passed in via arguments.
+            # Check to see whether any of these directories should be skipped..
+            dirs[:] = [d for d in dirs if pathlib.Path(root, d) not in self.skipped_dirs]
             for index, dir in enumerate(dirs):
```
Comment (spbrogan) [inline edk2toolext/environment/self_describing_environment.py:57]: Should the new path check be done inside the existing for loop?  Why do two iterations?

File: edk2toolext/environment/self_describing_environment.py [hunk 2]
```diff
@@ -30,6 +33,9 @@ def __init__(self, workspace_path, scopes=()):
         # Start with the provided set.
         self.scopes = scopes
 
+        # Allow certain directories to be skipped
+        self.skipped_dirs = tuple(map(pathlib.Path, (os.path.join(self.workspace, d) for d in skipped_dirs)))
```
Comment (spbrogan) [inline edk2toolext/environment/self_describing_environment.py:37]: should it be only workspace related or should packages path be involved?

File: edk2toolext/edk2_invocable.py [hunk 3]
```diff
@@ -61,6 +61,10 @@ def RetrieveCommandLineOptions(self, args: object) -> None:
         '''  Implement in subclass to retrieve command line options from the argparser namespace '''
         pass
 
+    def GetSkippedDirectories(self) -> Tuple[str]:
```
Comment (spbrogan) [inline edk2toolext/edk2_invocable.py:64]: this is implemented in base_abstract_invocable class.  Why needed here for same implementation?

File: edk2toolext/base_abstract_invocable.py [hunk 4]
```diff
@@ -33,6 +33,10 @@ def GetActiveScopes(self):
         ''' return tuple containing scopes that should be active for this process '''
         raise NotImplementedError()
 
+    def GetSkippedDirectories(self):
+        ''' return tuple containing workspace-relative directory paths that should be skipped for processing'''
```
Comment (spbrogan) [inline edk2toolext/base_abstract_invocable.py:37]: should we support packages path too?

File: edk2toolext/invocables/edk2_update.py [hunk 5]
```diff
@@ -61,6 +62,10 @@ def RetrieveCommandLineOptions(self, args):
         '''  Retrieve command line options from the argparser '''
         super().RetrieveCommandLineOptions(args)
 
+    def GetSkippedDirectories(self):
```
Comment (spbrogan) [inline edk2toolext/invocables/edk2_update.py:65]: why is this needed if the super has the function?

File: edk2toolext/edk2_invocable.py (hunk 3 above)
Comment (makubacki) [reply edk2toolext/edk2_invocable.py:64]: I just forgot to remove after adding it in the abstract class. I'll update it to be removed.

File: edk2toolext/invocables/edk2_update.py (hunk 5 above)
Comment (makubacki) [reply edk2toolext/invocables/edk2_update.py:65]: Added when debugging something and I forgot to remove it. I'll update it.

File: edk2toolext/environment/self_describing_environment.py (hunk 2 above)
Comment (makubacki) [reply edk2toolext/environment/self_describing_environment.py:37]: I was only considering workspace relative paths. Do you think packages path is needed?

**Comment by spbrogan** [comment]
also note it looks like pr_eval on project mu is broken.  is that expected?

**Review by corthon — APPROVED** [review]
Approved with agreement with Sean.

**Comment by makubacki** [comment]
> also note it looks like pr_eval on project mu is broken. is that expected?

Not expected, it was working in the previous push. It looks like GetSkippedDirectories() being removed from Edk2InvocableSettingsInterface() introduced it. I added it back to confirm.

File: edk2toolext/base_abstract_invocable.py (hunk 4 above)
Comment (makubacki) [reply edk2toolext/base_abstract_invocable.py]: I included a note about absolute paths.

**Review by spbrogan — APPROVED** [review]
